src/callgraphwidget.cpp

```diff
@@ -0,0 +1,208 @@
+        generateCallgraph(m_currentSymbol);
+    });
+
+    m_graphview = service->createInstance<KParts::ReadOnlyPart>(this);
+    if (!m_graphview) {
+        return;
+    }
+
+    layout->addWidget(costThreshold);
+    layout->addWidget(m_graphview->widget());
+
+    m_interface = qobject_cast<KGraphViewer::KGraphViewerInterface*>(m_graphview);
+
+    updateColors();
+    m_interface->setLayoutMethod(KGraphViewer::KGraphViewerInterface::LayoutMethod::InternalLibrary);
+
+    connect(m_graphview, SIGNAL(hoverEnter(QString)), this, SLOT(hoverEnter(QString)));
+    connect(m_graphview, SIGNAL(hoverLeave(QString)), this, SLOT(hoverLeave(QString)));
+
+    // used to find mouseclicks in the kgraphviewer part
+    qApp->installEventFilter(this);
+
+    m_graphFile->open();
+
+    connect(Settings::instance(), &Settings::callgraphChanged, this, [this]{
+        generateCallgraph(m_currentSymbol);
+    });
+}
+
+void CallgraphWidget::selectSymbol(const Data::Symbol& symbol)
+{
+    if (symbol == m_currentSymbol) {
+        return;
+    }
+
+    generateCallgraph(symbol);
+}
+
+void CallgraphWidget::hoverEnter(const QString& node)
+{
+    m_currentNode = node;
+}
+
+void CallgraphWidget::hoverLeave(const QString& node)
+{
+    Q_UNUSED(node);
+    m_currentNode = QString();
+}
+
+bool CallgraphWidget::eventFilter(QObject* watched, QEvent* event)
+{
+    Q_UNUSED(watched);
+    if (!isVisible())
+        return false;
+
+    if (event->type() == QEvent::MouseButtonPress) {
+        const auto e = static_cast<QMouseEvent*>(event);
+
+        if (m_graphview->widget()->geometry().contains(e->pos())) {
+            if (e->button() == Qt::MouseButton::LeftButton && !m_currentNode.isEmpty()) {
+                if (m_nodeIdToIndex.contains(m_currentNode)) {
+                    emit callerCalleeIndexChanged(m_nodeIdToIndex[m_currentNode]);
+                    m_currentNode.clear();
+                }
+            }
+        }
+    }
+    return false;
+}
+
+void CallgraphWidget::changeEvent(QEvent* event)
+{
+    QWidget::changeEvent(event);
+
+    if (event->type() == QEvent::PaletteChange) {
+        updateColors();
+    }
+}
+
+void CallgraphWidget::showEvent(QShowEvent* event)
+{
+    if (!m_graphview) {
+        return;
+    }
+
+    if (!event->spontaneous()) {
+        m_graphview->openUrl(QUrl(QStringLiteral("file://") + m_graphFile->fileName()));
+    }
+}
+
+void CallgraphWidget::generateCallgraph(const Data::Symbol& symbol)
+{
+    if (!m_graphview) {
+        return;
+    }
+
+    m_graphFile->resize(0);
+    m_currentSymbol = symbol;
+
+    QTextStream stream(m_graphFile);
+    writeGraph(stream, symbol);
+    stream.flush();
+
+    // if openUrl is called before the window is open it will freeze the application
+    if (isVisible()) {
+        m_graphview->openUrl(QUrl(QStringLiteral("file://") + m_graphFile->fileName()));
+    }
+}
+
+void CallgraphWidget::writeGraph(QTextStream& stream, const Data::Symbol& symbol)
```

Comment:
most of this function should be hoisted out of the widget and made reusable, such that we can test the dot graph generation completely without being tied to a widget